File: iOS/LeetCodeSwift/LeetCodeSwift/Leetcode/Merge_Two_Sorted_lists.swift
//
//  Merge_Two_Sorted_lists.swift
//  LeetCodeSwift
//
//  Created by tree on 2018/3/13.
//  Copyright © 2018年 treee. All rights reserved.
//

import Foundation

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

extension Solution {
    func mergeTwoLists(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        if l1 == nil && l2 ==  nil { return nil }
        guard let list1 = l1 else { return l2 }
        guard let list2 = l2 else { return l1 }
        var mergeNode: ListNode!
        if list1.val < list2.val {
            mergeNode = list1
            mergeNode.next = mergeTwoLists(list1.next, list2)
        }else {
            mergeNode = list2
            mergeNode.next = mergeTwoLists(list1, list2.next)
        }
        return mergeNode
    }
}

File: python/LeetCode/Merge_Two_Sorted_Lists.py
from Structs import ListNode

"""
合并两个链表，
e.g:
Input: 1->2->4, 1->3->4
Output: 1->1->2->3->4->4

思路：值传递，递归
"""
class Solution(object):

    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        # is 用来判断a，b对象是否就是一个对象；
        # == 判断a，b两个对象的值是否相等，是通过其value来判断的
        if l1 is None and l2 is None:
            return None
        mergeNode = None
        if l1.val < l2.val:
            mergeNode = l1
            mergeNode.next = self.mergeTwoLists(l1.next, l2)
        else:
            mergeNode = l2
            mergeNode.next = self.mergeTwoLists(l1, l2.next)
        return mergeNode